libgnucash/engine/qofchoice.cpp

```diff
@@ -21,107 +21,96 @@
  *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
  */
 
+#include <algorithm>
+#include <numeric>
+#include <vector>
+#include <string>
+#include <unordered_map>
 #include <config.h>
-#include <glib.h>
 
 #include "qof.h"
 #include "qofchoice.h"
 
+using OptionVec = std::vector<char*>;
+using ParamMap = std::unordered_map<std::string, OptionVec>;
+using ChoiceMap = std::unordered_map<std::string, ParamMap>;
 static QofLogModule log_module = QOF_MOD_CHOICE;
-static GHashTable *qof_choice_table = NULL;
-
-/* To initialise, call qof_choice_add_class in
-qof_object_register for the choice object. */
-static gboolean qof_choice_is_initialized(void)
-{
-    if (!qof_choice_table)
-    {
-        qof_choice_table = g_hash_table_new(g_str_hash, g_str_equal);
-    }
-    if (!qof_choice_table)
-    {
-        return FALSE;
-    }
-    return TRUE;
-}
+static ChoiceMap qof_choice_table;
 
 gboolean qof_object_is_choice(QofIdTypeConst type)
 {
-    gpointer value = NULL;
+    g_return_val_if_fail(type != NULL, false);
 
-    if (!qof_choice_is_initialized())
+    auto choice_it = qof_choice_table.find (type);
+    if (choice_it != qof_choice_table.end())
     {
-        return FALSE;
-    }
-    g_return_val_if_fail(type != NULL, FALSE);
-    value = g_hash_table_lookup(qof_choice_table, type);
-    if ((GHashTable*)value)
-    {
-        return TRUE;
+        return true;
     }
     DEBUG (" QOF_TYPE_CHOICE setup failed for %s\n", type);
-    return FALSE;
+    return false;
 }
 
 gboolean
 qof_choice_create(char* type)
 {
-    GHashTable *param_table;
-
-    g_return_val_if_fail(type != NULL, FALSE);
-    g_return_val_if_fail(qof_choice_is_initialized() == TRUE, FALSE);
-    param_table = g_hash_table_new(g_str_hash, g_str_equal);
-    g_hash_table_insert(qof_choice_table, type, param_table);
+    g_return_val_if_fail(type != NULL, false);
+    qof_choice_table.insert_or_assign (type, ParamMap{});
     return TRUE;
```

Comment:
Be consistent: Either always return the macro or always return the C++ bool value. IMO the macro is better because it doesn't require the implicit cast.